File: csharp/luhn/Luhn.cs
using System;
using System.Linq;

public static class Luhn
{
    public static bool IsValid(string number)
    {
        try
        {
            var results =
                number.Where(c => c != ' ').Reverse()
                .Select((c, i) => 
                {
                    if (!char.IsNumber(c))
                        throw new ArgumentException();
                    var value = int.Parse(c.ToString());
                    if(i % 2 == 1)
                        return value > 4 ? value * 2 - 9 : value * 2;
                    else 
                        return value;
                }).ToArray();
            return results.Length > 1 && results.Sum() % 10 == 0;
        }
        catch (ArgumentException)
        {
            return false;
        }
    }
}

File: go/luhn/luhn.go
package luhn

import (
	"unicode"
)

// Valid tests if the input is a valid luhn number
func Valid(input string) bool {
	var offset, sum, skip, count int
	if len(input)%2 == 1 {
		offset = 1
	}

	for i, char := range input {
		if char == ' ' {
			skip++
			continue
		}

		if !unicode.IsNumber(char) {
			return false
		}

		toAdd := int(char - '0')

		if (i+offset+skip)%2 == 0 {
			toAdd *= 2
			if toAdd > 9 {
				toAdd -= 9
			}
		}

		sum += toAdd
		count++
	}

	return count > 1 && sum%10 == 0
}
